src/libsyntax/parse/parser.rs

```diff
@@ -2164,6 +2157,16 @@ impl<'a> Parser<'a> {
                     assert!(self.eat_keyword(kw::Try));
                     return self.parse_try_block(lo, attrs);
                 }
+
+                // Span::rust_2018() is somewhat expensive; don't get it repeatedly.
+                let is_span_rust_2018 = self.span.rust_2018();
+                if is_span_rust_2018 && self.check_keyword(kw::Async) {
+                    return if self.is_async_block() { // check for `async {` and `async move {`
+                        self.parse_async_block(attrs)
+                    } else {
+                        self.parse_lambda_expr(attrs)
+                    };
+                }
```

Comment:
Nit: Looks like async/await treatment can be moved even further to the end of if-else chain below.

Reply:
This code consists of multiple `if/return` blocks followed by a chain of `if/else` blocks. The `async` keyword is handled in the `if/return` blocks and I moved it as late as possible among those blocks. The `await` keyword is handled in the `if/else` blocks. It might be possible to put the two keywords together but I didn't want to change things around too much. In the common case none of these keywords will match and we'll fall through to `parse_literal_maybe_minus` at the bottom, in which case the keyword ordering doesn't matter. Avoiding calling `rust_2018` is the important part, and the patch does that.

More generally, I suspect this whole function could be improved significantly, so that we don't have to check for every possible keyword before parsing a literal. But that should be a follow-up.